spock-specs/src/test/groovy/org/spockframework/util/ReflectionUtilSpec.groovy

```diff
@@ -200,4 +205,116 @@ class ReflectionUtilSpec extends Specification {
   static class Generics<T> {
     void foo(T one, List<T> two, List<String> three) {}
   }
+
+  def "isAnnotationPresentRecursive"(Class<?> clazz, Class<? extends Annotation> annotation, boolean expectedResult) {
+    expect:
+    ReflectionUtil.isAnnotationPresentRecursive(clazz, annotation) == expectedResult
+    where:
+    clazz              | annotation | expectedResult
+    Object             | Nullable   | false
+    Specification      | Testable   | true
+    ReflectionUtilSpec | Testable   | true
+    ReflectionUtilSpec | Nullable   | false
+    TypeWithAnno       | Issue      | true
+    SubTypeNoAnno      | Issue      | true
+  }
+
+  def "getAnnotationRecursive"(Class<?> clazz, Class<? extends Annotation> annotationClass, boolean expectedResult) {
+    when:
+    def anno = ReflectionUtil.getAnnotationRecursive(clazz, annotationClass)
+    if (expectedResult) {
+      assert clazz && annotationClass && anno != null
+    } else {
+      assert clazz && annotationClass && anno == null
+    }
```

Comment:
I think it would be better to split this feature, so that we don't have to use if-else here, and it should be in the `then` block.